### ISSUE TITLE
Search returning out-of-bounds results

### ISSUE THREAD

**Issue by bryevdv** [StanfordLegion/prof-viewer#30]
When the viewport is zoomed into a smaller interval, and the search is invoked, then the returned results can include items outside the viewport, contrary to expectation. The root cause appears to be a missed cache invalidation for unexpanded results.

cc @lightsighter for a profile and search term to use to reproduce.

**Comment by lightsighter** [comment]
I've sent the data and the instructions offline since they are too big to upload here.

**Comment by elliottslaughter** [comment]
I think we fixed this with #32?

**Comment by bryevdv** [comment]
@elliottslaughter I don't think so? #32 just added the whole-word search option, which makes it easier to pare down search for certain kinds of results, e.g. ids that have fences like `<123>`

**Comment by elliottslaughter** [comment]
You're right. The cache invalidation bug is still there. Sorry about the noise.

**Comment by bryevdv** [comment]
Hi @elliottslaughter @lightsighter I want to make sure I understand this bug. 

* Starting all the way zoomed out, see an item with "71"

  <img width="354" alt="Screenshot 2023-11-13 at 13 54 27" src="https://github.com/StanfordLegion/prof-viewer/assets/1078448/577fa5d7-0a3d-4315-929e-0f4b3ca1ba91">

* Search for "71" (still zoomed out) see the result, all good:

  <img width="349" alt="Screenshot 2023-11-13 at 13 54 42" src="https://github.com/StanfordLegion/prof-viewer/assets/1078448/a46f7c6c-da6a-484e-b63f-1901d65600bb">

* Reset the search, zoom into a region *without* the "71" then search again &mdash; no result shown, *still all good?*

  <img width="349" alt="Screenshot 2023-11-13 at 13 55 00" src="https://github.com/StanfordLegion/prof-viewer/assets/1078448/7be10145-cd1e-4513-8edf-9a8945ff3477">

  I had thought this was the reported error case, and expected to see a result show up even when searching in the zoomed-away area. But it seems to be working as I'd expect.

---

By contrast: If I start from the zoomed out view with a search result (bullet two above), and *then* zoom away, without resetting anything first, then the result stays in the search, which I would *not* expect:

  <img width="354" alt="Screenshot 2023-11-13 at 13 55 33" src="https://github.com/StanfordLegion/prof-viewer/assets/1078448/5575b740-943a-432b-b725-b29eb33fd078">

Is this different from the original reported bug? Regardless, just to confirm: the search results should reset when the view interval changes?

**Comment by lightsighter** [comment]
> Reset the search, zoom into a region without the "71" then search again — no result shown, still all good?

If I scroll right or left, will the results automatically refresh?

>  I had thought this was the reported error case, and expected to see a result show up even when searching in the zoomed-away area. But it seems to be working as I'd expect.

I honestly don't remember what I was trying to do that was behaving funny. I just remembered you guys agreed with me that it was strange when I showed it to you. 😇

> By contrast: If I start from the zoomed out view with a search result (bullet two above), and then zoom away, without resetting anything first, then the result stays in the search, which I would not expect:

By "zoom away" do you mean "zoom in"? If so I think this might have been my case too. I was trying to refine a search and wanted to see results getting pruned out as the window size shrunk.

> Regardless, just to confirm: the search results should reset when the view interval changes?

I think in general this in the invariant we want to maintain: the search results should always reflect what is visible. Anything that causes the viewport to change, should cause the search results to refresh.

**Comment by elliottslaughter** [comment]
I think the bug involved a combination of zooming + the "search unexpanded processors" option. Because those processors are unexpanded, the cache invalidation logic doesn't kick in and we get stale results. I think that's why I made my comment about `inflate_meta` here: https://github.com/StanfordLegion/prof-viewer/issues/30#issuecomment-1758090133

Try this in a fresh profile:

 1. Find the task you want to search for
 2. Collapse all the processors (so you can't see it anymore)
 3. Search (should be empty; this is correct)
 4. Click "search unexpanded processors"
 5. It should appear in search now (this is correct)
 6. Zoom away (anywhere that the task isn't, so the task isn't in the view interval)
 7. The task should still appear in the search (**this is the bug**)

**Comment by elliottslaughter** [comment]
A fix is (very belatedly) in #81.
